rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,430 @@
+
+This solves two problems: 1) multiple users with on the same remote machine will not have overlapping credentials on disk in `~/.tsh` and 2) attackers will not be able to steal a user's credentials from disk.
+
+2. **Issued certificates through headless authentication are short lived (1 minute TTL).**
+
+Since the user certificates are only meant to last for a single client lifetime, the Auth server should issue the certificates with a 1 minute TTL. This limits the damage potential if an attacker manages to compromise the user's certificate, whether through a faulty client or a clever attack.
+
+3. **User must complete WebAuthn authentication for each individual headless `tsh` request (`tsh ls`, `tsh ssh`, etc.).**
+
+Headless authentication, like any login mechanism, can be started by any unauthenticated user. To prevent phishing attacks, we must prompt the user to acknowledge and approve each headless authentication request with WebAuthn, since WebAuthn provides strong protection against phishing attacks. Legacy OTP MFA methods will not be supported.
+
+4. **The Server must verify that the client requesting certificates is the only client that can use the certificates.**
+
+Like other login procedures, this be accomplished with a PKCE-like flow, where the client provides their public key in the login request, and the server issues certificates for the client's public key. If the client does not have access to the corresponding client private key, they cannot use the certificates.
+
+Note: Ideally, we would ensure that the only client that can *receive* certificates is the requesting client, but ensuring the PKCE flow above is enough to defend against attacks since the certificates are useless without the client's private key. In the current design this is accomplished by encapsulating the headless login flow between the remote machine and Teleport into a single endpoint - `/webapi/login/headless`
+
+5. **(optional) limit the scope of headless certificates**
+
+When a user requests headless login certificates, we know exactly what command they are trying to complete. In theory, this means that we can scope their certificate to provide the most limited possible privileges necessary to complete said action. This would help to reduce the blast radius if an attacker manages to steal the user's headless certificates.
+
+However, unlike Per-session MFA which scopes certificates to a specific node connection, we would need to scope the certificates to the entire API flow for a `tsh` command. For example, `tsh ssh` needs to:
+
+1. Connect to Proxy
+2. List nodes
+3. View cluster details (Auth preference, etc.)
+4. Connect to Node
+
+Given the implementation complexity, this is likely overkill as we've already reduced this attack vector significantly:
```

Comment:
Really cool idea, and at first I was going to agree with you that it may not be worth the complexity.  However, after more thought I wonder if we do want to consider some mechanism like this.  

When considering an attack where the `tsh` client has been replaced with a malicious client, I don't currently see any enforcement to control that what the user has approved in the web UI is exactly what's allowed to execute.

Is there validation of what execution is possible outside of client logic?

Reply:
Agreed, see the above thread about this concern - https://github.com/gravitational/teleport/pull/21005#discussion_r1096322566

I'm adding details for limiting the certificate privileges, but it adds a lot of complexity and further security questions/concerns. Let me know what you think once I've pushed the RFD changes.